### ISSUE TITLE
Make is possible to install apps from the Store 

### ISSUE THREAD

**Issue by kelps** [Microsoft/windows-dev-box-setup-scripts#30]
There should be a way to improve the scripts to make it possible to install apps directly from the Store. Example: when I repave my dev machines, i install some apps like "Open Live Writer", Paint.NET, Spotify, some Edge extensions e a few other Store apps I use regularly.

It looks like the current scripts already install some appx files that supposedly come from the store, but the process it uses to do so seems to be a hack. It downloads the appx from a minified shortened url and then sideloads it. As far as I could tell, there is no way to get a fixed url to the appx files I want from the Store and do the same, as the urls have information as version in them. I also don't know if installing an app this way has any side effects on how the app is updated from the Store from that point forward, if at all, since it was sideloaded.

The optimal option would be to call some official Windows API passing the Store ID for the app we want to install and let it do it's thing by installing from the Store instead of sideloading. If the user doesn't already "own" the app, it could ask to confirm the purchase (even if free) or have a flag to do it if necessary (similar to the "-y" in Chocolatey).

**Comment by yodurr** [comment]
Love this feedback, thank you.

**Comment by DarthSpock** [comment]
So in Windows proper we do have this Powershell command (The example comes from https://docs.microsoft.com/en-us/windows/wsl/install-on-server):

`Invoke-WebRequest -Uri https://aka.ms/wsl-ubuntu-1604 -OutFile ~/Ubuntu.zip -UseBasicParsing`

We also have `curl` and `tar` available in Windows now. Don't see `tar` being very useful other than for Linux distros (or any tar files that somehow make it into the Store)  but `curl` does sorta replace (but not entirely) `Invoke-WebRequest`

**Comment by kelps** [comment]
@DarthSpock this doesn't address the issue I presented. This command is simply downloading something from somewhere. I want to install something from the Store, preferably using official Store APIs in my computer.

The .ps1 scripts here use this to download the .appx file from somewhere and then sideload them. My concerns with this are twofold:

1. The .appx download origin is not the Store (maybe in the files in the script it is the same file, but it is not coming from the same url the Store uses) 
2. The .appx is being sideloaded. I'm not sure, but I think it will not get updated if the developer updates it in the Store. (please correct me if I'm wrong)

How do I install, for example, "Paint.NET" from the Store using scripting? Is it possible to generate an official download url for a Store .appx and use the combination `Invoke-WebRequest` + `Add-AppxPackage` to install it and still get updates? Do I need to know the app version in the Store to get the url or can I get a "default" url for the app that will always point to the latest version? How would I generate such urls for the apps I want/need to install when I repave?

An even better solution would be to have something like `Acquire-StorePackage`, where you would pass the app name, key or guid (or something) and it would install it from the Store. I understand that this request is more for the Windows Store team than for the champions in this repo, but I suppose the relevant people might see this and take pity on us folks trying to automate things...

**Comment by kelps** [comment]
Reading some documentation while replying to @DarthSpock's comment, I found a few articles using a combination of `Get-AppxPackage` + `Add-AppxPackage` to reinstall apps the current user removed or that got "lost" in some Insiders update. It got me thinking:

If I use ´Get-AppxPackage` on my machine before repaving it to get and save the manifest files for the apps I will want to get back after the repave, it might be possible to use `Add-AppxPackage` later pointing to the saved manifests and have the apps reinstalled. I'll have to investigate that on my next repave (might be time to do it on my old Surface Pro 2 anyway). If I find the time over the weekend, I'll let you know of my findings.

I'm now also thinking of using `Remove-AppxPackage` in my script to remove some unwanted apps that Microsoft now insists on pushing to users (I am looking at you Candy Crush...)

**Comment by DarthSpock** [comment]
So the https://aka.ms/wsl-ubuntu-1604 is from the Store but automatically provides the option to install the appx without the need to go through the Store to get it. As far as getting updates goes, well if you're in a corporate network, you probably won't get them if you don't have Store access. Which makes standalone appx for different apps extremely useful. In larger organizations, there is typically a workstation that will have access in order to obtain certain apps and will go through a test, then update phase. We can't assume all devs will have access to the Store so while this may not fit your specific needs, it will address somebody else's.

**Comment by gep13** [comment]
@kelps can you perhaps to the documentation that you were looking at?  Thanks!

**Comment by gep13** [comment]
@kelps thanks for providing those links, appreciate it!

**Comment by kelps** [comment]
@DarthSpock Yep, I understand and agree this is a valid approach, but catering only to big companies isn't how things should work anymore. Specially in the dev circles, we have more and more hobbyists, remote workers and BYOD friendly companies. 
What I am asking here (hoping actually) is for someone from Microsoft to acknowledge and address (or at least start the process of addressing) the need for APIs that will allow end users and IT managers to install apps from the Store in an automated way.
Of course, I also understand these kinds of APIs may be posed as a security threat and need to be properly implemented so as to not be abused by malicious actors (malware, viruses, rogue programs, hackers, disgruntled employees, bored/grounded kids, ...).
Anyway, hoping for a better automation story here and really excited by the mere existence of this repo. We are making progress ;)

**Comment by DarthSpock** [comment]
Yes, I think this project needs to address both scenarios. Those who can use the Store and those can't. I'm still looking forward to the day when we can honestly get all dev work done using a combo of WSL and Windows but unfortunately the dev community targetted doesn't include anything GUI or hardware related if it's for Linux, which is a shame in my mind. If they could get AMD/NVIDIA/Intel onto WSL, well now, they would be able to bring in far more profit and the rest of us would have the advantage of fully harnessing our hardware for WSL (thus drastically improving performance).

**Comment by kelps** [comment]
I am almost sure I had already added this feedback using the Feedback Hub app a long time ago (certainly more than a year).

I didn't find it, so I created a new one in the following link. Upvotes are appreciated. [Provide a way (scripting api/commands) for users to install/remove Store apps via command-line scripts](https://aka.ms/AA1ajjk)

**Comment by kelps** [comment]
@DarthSpock I think the apis should allow installing apps directly from the Store, no matter what kind of app it is, so not only WSL or a specific kind/category. It should behave with the same restrictions you would get when using the Store app, so if an app is blocked by a policy or unavailable to the user, it also would be blocked via scripts. Any powershell commands created for this should use the same code the Store already uses, to ensure all the same rules apply.

I would also add some security features like: by default the command only installs apps the user already owns or apps that are free (there should be argument flags to override this behavior, of course). If the user doesn't own the app, the Store purchase UI should be displayed to confirme the purchase (there could be a flag that dispenses this step if the app is free, but it shouldn't be on by default).

**Comment by DarthSpock** [comment]
@kelps Nothing you said is opposite of what I was saying. I'm in full agreement, there should be APIs for installing from the store. But if those APIs require connectivity to the Store itself, it will most likely be blocked on most corporate networks. But it also might be an easier sell to corporations to allow API connectivity for command-line only access vice giving users Store access (not to mention no admin account should have web access whatsoever including the API connectivity). This access could then be created as a security group in AD that only those who require it would be granted access to it.

**Comment by DarthSpock** [comment]
Also a note to bear in mind for future: there is a initiative to make apps cross-platform that would be hosted in the Store. In essence, a one and done for developers who have to target multiple platforms. This is still early in it's infancy but eventually this project will need to address those APIs for connection to other platforms besides Windows.

**Comment by DarthSpock** [comment]
> If the user doesn't own the app, the Store purchase UI should be displayed to confirm the purchase (there could be a flag that dispenses this step if the app is free, but it shouldn't be on by default)

To add to that, if the app not owned is a free app or has a free trial, that will need to be addressed as well. I'm also seeing a logistic and security issue here. If in a corporate environment, the developer won't have authorization to purchase an app from the Store. It would have to go through their finance and security first. A mechanism will need to exist to ensure that unauthorized apps aren't installed (that prevents Group Policy/HIPS, etc. from kicking off). It wouldn't be fair to the business to have to put safe rails on something they aren't responsible for that they now need to dedicate resources towards if they want to allow the connection.

**Comment by bitcrazed** [comment]
@kelps: I'd suggest creating & sharing a new feedback item using the same text, but targeting the "Microsoft Store" | "Installing Apps" categories. This isn't really about payment per se (though that will have to be part of the equation for any solution).

![image](https://user-images.githubusercontent.com/961950/40087806-0db33d84-5859-11e8-92e1-809193e22bee.png)

**Comment by kelps** [comment]
@bitcrazed Ohhhh, That was the intent. I must've slipped or scrolled to another sub category without noticing. Thanks for pointing it out. I just created a new feedback in the correct sub-category. I'll post the link here as soon as it shows up in the "My Feedback" tab.

**Comment by kelps** [comment]
@DarthSpock I agree with all you said. If it sounded like I was disagreeing, I apologize.

**Comment by RobCannon** [comment]
@DarthSpock I am very interested in that, but that won't be viable until the next release of Windows 10 for support of the .MSIX format.

**Comment by DarthSpock** [comment]
True, it would only help those on the latest Win 10 but does have the potential for resolving this issue at least for that build and up and of course recommend upgrading to that build. But yes, we'll need something else in the mean-time to support earlier versions of Win 10 along with 7 and 8.1.

**Comment by kelps** [comment]
To me, it looks like MSIX is tool mainly for enterprises and software vendors/developers. It doesn't look like it'll help people that simply want to script the installation of software from the Microsoft Store on Windows 10+. 

I'm not saying MSIX is bad. It is a HUGE step forward for software on Windows and I can't wait for vendors to start making their app installers available in this format, as it'll guarantee computer stability, app isolation and auto-update. MSIX is actually a perfect format for stuff to be delivered using Chocolatey. And, if I'm not mistaken, there is a KB that adds support for MSIX on Windows 7. Of course app isolation and some of the other aspects will be Win10+ only, but this makes it even more useful.

**Comment by bitcrazed** [comment]
@kelps: Thanks for the feedback and suggestions.

Alas, the point of these scripts is to automate the setup of a newly-installed/re-paved dev box. 

Since Windows Store doesn't yet provide an app installation automation mechanism, automating the installation of store apps is sub-optimal at best, and breaks our automation goals at best :(

However, we do continue to discuss with the store team. Hoping we'll be able to use the feedback like the above as fuel for these discussions ;)

And, yes, MSIX, doesn't directly benefit the install automation story for end-users, but it does provide benefits wherein if app vendors publish their apps in MSIX packages, it'll make the automation job a little easier and more consistent.

**Comment by kelps** [comment]
I just created a repo to store my repave related scripts [repave-scripts](https://github.com/kelps/repave-scripts). There I have the script I am currently using when I repave my PCs. I was storing those scripts in my OneDrive account, but I figured that having them in source control is a better solution. The code I mentioned above is there, if anyone is curious. 

I used that script recently to repave my work laptop and it worked quite well, but not as well as it would have if I could automate the Windows Store installs. I also wish Visual Studio and SQL Server Management Studio where available in the store, but I get why they are not (yet anyway) and in any case this has nothing to do with this issue or repo.

**Comment by bitcrazed** [comment]
Hey @Kelps. Thanks for sharing. A few notes:

1. If you were to rename your script to *.ps1, you wouldn't need to invoke PowerShell at the start
1. I recommend you begin and end your script by disabling and re-enabling UAC, etc.
1. Another benefit of moving to PowerShell is that you can explicitly modify your registry settings via `HKLM:...` or `HKCU:` registry mounts, further reducing your file count.
1. Visual Studio can be installed by Chocolatey and you can control which features you want enabled! See https://chocolatey.org/packages/visualstudio2017enterprise for more info
1. Same as above for SQL Management Studio: https://chocolatey.org/packages/sql-server-management-studio
1. You may also prefer to NOT pause between each package install so you can walk away and leave your machine to install everything as quickly as possible without requiring constant monitoring :)

HTH.

**Comment by kelps** [comment]
Thanks for the tips @bitcrazed.

1- I know I could use .ps1 files for this instead of .bat, but I prefer the simplicity of being able to double click it to run. This was a conscious decision, but I might change it to a .ps1 file and have the .bat simply invoke PowerShell to run it.
2- I'll look into it. Do any of the scripts here have the code for disabling and enabling UAC? Is it doable from .bat or only .ps1?
3- Ok, that is definitely a benefit that I have not considered before. I do like the .reg files because I just had to export them, but having it in the script reduces the click count.
4- 5- This I didn't know and I love it! I'll have to try it. At least 2 manual installations I will not have to deal with.
6- I had the pauses there because those installations from chocolatey used to fail a lot when I started doing this. This made it easier for me to see that something went wrong, but I suppose you're right. I only really need the pauses now between the Windows Store apps (I really hope this will improve eventually)

Once again, thanks for the tips.

**Comment by wireddown** [comment]
> Thanks for responding. FWIW, when I click your link to feedback hub I get, "Your account doesn't have access to this feedback."

Same for me :(

**Comment by VijayS1** [comment]
I have found a temporary solution by allowing the user to manually click the install button in the store app. see my [solution here](https://github.com/VijayS1/windows-dev-box-setup-scripts/blob/master/scripts/InstallStoreApps.ps1)

**Comment by ledgeruser2** [comment]
You can absolutely download the .appx files and their dependencies from the Microsoft Store via PowerShell in a completely automated fashion, as long as the app is already installed on your computer. It's not because you need the package contents, it's because you need the app info to force the Store to re-download the .appx files and keep them permanently.
Installing a new app from PowerShell from the store is a bit tricky. I am currently working on it.

**Comment by WSLUser** [comment]
Winget would most likely be the best way forward with this

**Comment by manhinhang** [comment]
I created CLI tools for download app from store but it required start `Appium` server locally to automate stuff.
https://github.com/manhinhang/ms-store-install-cli

**Comment by bitcrazed** [comment]
@manhinhang Appreciate your effort here, but driving the Microsoft Store via UI automation is likely to be very error prone and fragile.

As @WSLUser and others have stated and as I can confirm, installation via winget will be the preferred approach. Appreciate everyone's patience, but it'll be worth the wait ;)

**Comment by Floccinaucinihilipilification11** [comment]
@bitcrazed You mean, about as error-prone as your average piece of Microsoft software? 🤣🤣🤣

**Comment by bitcrazed** [comment]
@Floccinaucinihilipilification11 No, even worse than that!
<img src="https://media.giphy.com/media/5MUnOfrQwFcdO/giphy.gif"/>

**Comment by Floccinaucinihilipilification11** [comment]
Searching the MS Store from PowerShell, First Beta

https://dev.to/ledgeruser2/search-microsoft-store-from-powershell-2bjj
